File: macos-swiftui/Sources/Core/CLIVersion.swift
/* The repository's CLI_VERSION pin.
 *
 * Plain KEY=VALUE, deliberately, so bash, PowerShell, a C test and this can all
 * read it without a library. Parsing it here rather than in the test file is
 * what lets the Health pane show the same numbers the conformance suite
 * asserts on.
 *
 * REQUIRES_ARCHIVE_LAYOUT and `supportedArchiveLayout` in Archive.swift are two
 * copies of one fact. The pin is what an installer and a human read; the
 * constant is what actually decides, per video, whether a folder is rendered or
 * flagged. Them disagreeing is how an app ends up claiming to read a layout it
 * does not, so the suite asserts they match.
 */

import Foundation

struct CLIVersion {
    var values: [String: String] = [:]

    var cliRepo: String? { values["CLI_REPO"] }
    var cliRef: String? { values["CLI_REF"] }
    var requiresArchiveLayout: Int64? {
        values["REQUIRES_ARCHIVE_LAYOUT"].flatMap { Int64($0) }
    }

    static func parse(_ text: String) -> CLIVersion {
        var v = CLIVersion()
        for rawLine in text.components(separatedBy: "\n") {
            let line = rawLine.trimmingCharacters(in: .whitespaces)
            if line.isEmpty || line.hasPrefix("#") { continue }
            guard let eq = line.firstIndex(of: "=") else { continue }
            let key = String(line[line.startIndex..<eq]).trimmingCharacters(in: .whitespaces)
            let value = String(line[line.index(after: eq)...]).trimmingCharacters(in: .whitespaces)
            if !key.isEmpty { v.values[key] = value }
        }
        return v
    }

    static func load(atPath path: String) -> CLIVersion? {
        guard let text = try? String(contentsOfFile: path, encoding: .utf8) else { return nil }
        return parse(text)
    }
}

File: windows-winui/YtdlWin/Core/CliVersion.cs
/* The repository's CLI_VERSION pin.
 *
 * Plain KEY=VALUE, deliberately, so bash, PowerShell, a C test, a Swift test
 * and this can all read it without a library. Parsing it here rather than in
 * the test file is what lets the Health pane show the same numbers the
 * conformance suite asserts on.
 *
 * REQUIRES_ARCHIVE_LAYOUT and Archive.SupportedArchiveLayout are two copies of
 * one fact. The pin is what an installer and a human read; the constant is what
 * actually decides, per video, whether a folder is rendered or flagged. Them
 * disagreeing is how an app ends up claiming to read a layout it does not, so
 * the suite asserts they match.
 */

using System;
using System.Collections.Generic;
using System.Globalization;

namespace YtdlWin.Core;

public sealed class CliVersion
{
    public Dictionary<string, string> Values { get; } = new(StringComparer.Ordinal);

    public string? CliRepo => Values.TryGetValue("CLI_REPO", out var v) ? v : null;
    public string? CliRef => Values.TryGetValue("CLI_REF", out var v) ? v : null;

    public long? RequiresArchiveLayout
    {
        get
        {
            if (!Values.TryGetValue("REQUIRES_ARCHIVE_LAYOUT", out var raw)) return null;
            return long.TryParse(raw, NumberStyles.Integer, CultureInfo.InvariantCulture, out var n)
                ? n : null;
        }
    }

    public static CliVersion Parse(string text)
    {
        var v = new CliVersion();
        // Split on '\n' after normalising, so a file checked out with CRLF
        // endings -- which on Windows is the default -- does not leave a '\r'
        // on the end of every value. "main\r" is not "main", and a pin that
        // compares unequal to itself is a bug that only appears on this
        // platform.
        var normalised = text.Replace("\r\n", "\n").Replace('\r', '\n');
        foreach (var rawLine in normalised.Split('\n'))
        {
            var line = rawLine.Trim();
            if (line.Length == 0 || line.StartsWith('#')) continue;

            var eq = line.IndexOf('=');
            if (eq < 0) continue;

            var key = line[..eq].Trim();
            var value = line[(eq + 1)..].Trim();
            if (key.Length > 0) v.Values[key] = value;
        }
        return v;
    }

    public static CliVersion? Load(string path)
    {
        var text = Paths.ReadAllTextOrNull(path);
        return text is null ? null : Parse(text);
    }
}
